src/plugins/gs-plugin-external-apps.c

```diff
@@ -0,0 +1,712 @@
+/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
+ *
+ * Copyright (C) 2016 Endless Mobile, Inc
+ *
+ * Author: Joaquim Rocha <jrocha@endlessm.com>
+ *
+ * Licensed under the GNU General Public License Version 2
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <config.h>
+#include <json-glib/json-glib.h>
+#include <flatpak.h>
+#include <glib.h>
+#include <gnome-software.h>
+#include <string.h>
+
+#include "gs-flatpak.h"
+
+#define EXTERNAL_ASSETS_SPEC_VERSION 1
+#define JSON_SPEC_KEY "spec"
+#define JSON_RUNTIME_KEY "runtime"
+#define JSON_RUNTIME_NAME_KEY "name"
+#define JSON_RUNTIME_URL_KEY "url"
+
+#define METADATA_URL "GnomeSoftware::external-app::url"
+#define METADATA_HEADLESS_APP "GnomeSoftware::external-app::headless-app"
+#define METADATA_BUILD_DIR "GnomeSoftware::external-app::build-dir"
+#define METADATA_EXTERNAL_ASSETS "flatpack-3rdparty::external-assets"
+
+struct GsPluginData {
+	GsFlatpak	*flatpak;
+};
+
+void
+gs_plugin_initialize (GsPlugin *plugin)
+{
+	GsPluginData *priv = gs_plugin_alloc_data (plugin, sizeof(GsPluginData));
+
+	priv->flatpak = gs_flatpak_new (plugin, GS_FLATPAK_SCOPE_USER);
+
+	/* Run plugin after the flatpak ones because we need them to install
+	 * the app's headless part first */
+	gs_plugin_add_rule (plugin, GS_PLUGIN_RULE_RUN_BEFORE, "flatpak-system");
+	gs_plugin_add_rule (plugin, GS_PLUGIN_RULE_RUN_BEFORE, "flatpak-user");
+}
+
+static gboolean
+gs_app_is_flatpak (GsApp *app)
+{
+	const gchar *id = gs_app_get_id (app);
+	return id != NULL &&
+		(g_str_has_prefix (id, GS_FLATPAK_USER_PREFIX ":") ||
+		 g_str_has_prefix (id, GS_FLATPAK_SYSTEM_PREFIX ":"));
+}
+
+void
+gs_plugin_adopt_app (GsPlugin *plugin, GsApp *app)
+{
+	const gchar *id = gs_app_get_id (app);
```

Comment:
You could use `gs_app_get_id()` directly in the `g_debug()` since it's the only place it's used and the string is const